### PR TITLE
[WIP] Implement project for Transform. #264

### PR DESCRIPTION
**Pull request by marvinlanhenke**
#264 
I just wanted to share this as a starting point - and to get some feedback, thoughts about the approach here.

@liurenjie1024 @Xuanwo 
Please take a look if this is a viable solution at all. 
I'm quite unsure about the `fn transform` and the handling of the arrow_array - seems kinda clunky?

Anyway, let me know what you think. best regards.

### CODE REVIEW

**Comment by marvinlanhenke** [comment]
@liurenjie1024 @Xuanwo 
the main design concerns I want to discuss:

1. exposing `op` and `literals` from expressions. For prototyping I did this directly on the expression - it might be better to implement this on `BoundPredicate` itself and expose `op` via pub(crate) from the struct?
2. Once I receive a `PrimitiveLiteral` I need to convert it to an arrow_array in order to perform the transformation. The conversion to the arrow array could also be implemented on `PrimitiveLiteral` itself?

I would really appreciate your thoughts on this

File: crates/iceberg/src/spec/transform.rs [hunk 1]
```diff
@@ -261,6 +269,50 @@ impl Transform {
...
+                        PrimitiveLiteral::Int(v) => func
+                            .transform(Arc::new(Int32Array::from_value(v, 1)))?
+                            .as_any()
+                            .downcast_ref::<Int32Array>()
+                            .ok_or_else(|| Error::new(ErrorKind::Unexpected, "Failed to downcast"))?
+                            .value(0),
+                        PrimitiveLiteral::Long(v) => func
+                            .transform(Arc::new(Int64Array::from_value(v, 1)))?
+                            .as_any()
+                            .downcast_ref::<Int32Array>()
```
Comment (sdd) [inline crates/iceberg/src/spec/transform.rs:297]: Should this be `Int64Array`?

**Comment by marvinlanhenke** [comment]
@liurenjie1024 @sdd @Xuanwo @ZENOTME 
I went on and implemented `fn project` for `Transform::Bucket` with some design assumptions. PTAL and tell me what you think, before we can move on and implement the missing transforms.
Thank you so much for your effort

**Comment by ZENOTME** [comment]
Thanks for this job! 

> I'm quite unsure about the fn transform and the handling of the arrow_array - seems kinda clunky?

I think the transform can provide an interface like `transform_literal` later. 
```
pub trait TransformFunction: Send {
    /// transform will take an input array and transform it into a new array.
    /// The implementation of this function will need to check and downcast the input to specific
    /// type.
    fn transform(&self, input: ArrayRef) -> Result<ArrayRef>;
+ fn transform_literal(&self, literal:Literal) -> Result<Literal>; 
}
```

File: crates/iceberg/src/expr/predicate.rs [hunk 2]
```diff
@@ -187,6 +199,14 @@ impl<T> SetExpression<T> {
         debug_assert!(op.is_set());
         Self { op, term, literals }
     }
+
+    pub(crate) fn op(&self) -> PredicateOperator {
+        self.op
+    }
+
+    pub(crate) fn literals(&self) -> &FnvHashSet<Datum> {
```
Comment (liurenjie1024) [inline crates/iceberg/src/expr/predicate.rs:207]:
```suggestion
    pub(crate) fn literals(&self) -> impl Iterator<Item=Datum> {
```

How about returning iterator only? It's better not expose the inner data structure.

**Comment by liurenjie1024** [comment]
Hi, @marvinlanhenke This pr looks great to me. But I think @ZENOTME is right, we should implement #283 first. The reason we implement transform on arrow array could be found [here](https://github.com/apache/iceberg-rust/issues/32)

**Comment by ZENOTME** [comment]
> Hi, @marvinlanhenke This pr looks great to me. But I think @ZENOTME is right, we should implement #283 first. The reason we implement transform on arrow array could be found [here](https://github.com/apache/iceberg-rust/issues/32)

I will init the interface as soon as possible to avoid blocking this.

**Comment by marvinlanhenke** [comment]
> Hi, @marvinlanhenke This pr looks great to me. But I think @ZENOTME is right, we should implement #283 first. The reason we implement transform on arrow array could be found [here](https://github.com/apache/iceberg-rust/issues/32)

Thank you for providing the reason why we use arrow array in the first place.

**Comment by marvinlanhenke** [comment]
> > Hi, @marvinlanhenke This pr looks great to me. But I think @ZENOTME is right, we should implement #283 first. The reason we implement transform on arrow array could be found [here](https://github.com/apache/iceberg-rust/issues/32)
> 
> I will init the interface as soon as possible to avoid blocking this.

Sounds good to me

**Pull request closed without merging**